File: app/src/main/java/bassamalim/hidaya/activities/QuizActivity.java
package bassamalim.hidaya.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.TypedValue;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import androidx.appcompat.app.AppCompatActivity;
import androidx.room.Room;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import bassamalim.hidaya.R;
import bassamalim.hidaya.database.AppDatabase;
import bassamalim.hidaya.database.dbs.QuizAnswersDB;
import bassamalim.hidaya.database.dbs.QuizQuestionsDB;
import bassamalim.hidaya.databinding.ActivityQuizBinding;
import bassamalim.hidaya.other.Utils;

public class QuizActivity extends AppCompatActivity {

    private ActivityQuizBinding binding;
    private AppDatabase db;
    private List<QuizQuestionsDB> questions;
    private int current = 0;
    private final int[] cAnswers = new int[10];
    private Button nextBtn;
    private Button prevBtn;
    private RadioGroup radioGroup;
    private final RadioButton[] answerBtns = new RadioButton[4];
    private TypedValue colorText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.myOnActivityCreated(this);
        binding = ActivityQuizBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        db = Room.databaseBuilder(this, AppDatabase.class, "HidayaDB")
                .createFromAsset("databases/HidayaDB.db").allowMainThreadQueries().build();

        colorText = new TypedValue();
        getTheme().resolveAttribute(R.attr.myText, colorText, true);

        selectQuestions(getQuestions());

        Arrays.fill(cAnswers, -1);

        initViews();

        setListeners();

        ask(current);
    }

    private void selectQuestions(List<QuizQuestionsDB> rawQuestions) {
        Collections.shuffle(rawQuestions);
        questions = new ArrayList<>(rawQuestions.subList(0, 10));
    }

    private void initViews() {
        nextBtn = binding.nextQuestion;
        prevBtn = binding.previousQuestion;

        radioGroup = binding.answersRadioGroup;

        answerBtns[0] = binding.answer1;
        answerBtns[1] = binding.answer2;
        answerBtns[2] = binding.answer3;
        answerBtns[3] = binding.answer4;
    }

    private void setListeners() {
        for (int i = 0; i < answerBtns.length; i++) {
            int finalI = i;
            answerBtns[i].setOnClickListener(v -> answered(finalI));
        }

        prevBtn.setOnClickListener(v -> previousQ());
        nextBtn.setOnClickListener(v -> nextQ());
    }

    private void ask(int num) {
        QuizQuestionsDB q = questions.get(num);

        String qNum = getString(R.string.question) + " " + (current+1);
        binding.topBarTitle.setText(qNum);

        binding.questionScreen.setText(questions.get(current).getQuestion_text());

        List<QuizAnswersDB> answers = getAnswers(q.getQuestion_id());
        for (int i = 0; i < answerBtns.length; i++)
            answerBtns[i].setText(answers.get(i).getAnswer_text());

        adjustButtons();
    }

    private void adjustButtons() {
        radioGroup.clearCheck();
        if (cAnswers[current] != -1)
            radioGroup.check(answerBtns[cAnswers[current]].getId());

        if (current == 0) {
            prevBtn.setEnabled(false);
            prevBtn.setTextColor(getResources().getColor(R.color.grey, getTheme()));
        }
        else if (current == 9) {
            if (allAnswered()) {
                nextBtn.setText(getString(R.string.finish_quiz));
                nextBtn.setEnabled(true);
                nextBtn.setTextColor(colorText.data);
            }
            else {
                nextBtn.setText(getString(R.string.answer_all_questions));
                nextBtn.setEnabled(false);
                nextBtn.setTextColor(getResources().getColor(R.color.grey, getTheme()));
            }
        }
        else {
            prevBtn.setEnabled(true);
            prevBtn.setTextColor(colorText.data);

            nextBtn.setEnabled(true);
            nextBtn.setText(getString(R.string.next_question));
            nextBtn.setTextColor(colorText.data);
        }
    }

    private void answered(int a) {
        cAnswers[current] = a;

        adjustButtons();

        if (current != 9)
            nextQ();
    }

    private void nextQ() {
        if (current == 9)
            endQuiz();
        else
            ask(++current);
    }

    private void previousQ() {
        if (current > 0)
            ask(--current);
    }

    private boolean allAnswered() {
        for (int cAnswer : cAnswers) {
            if (cAnswer == -1)
                return false;
        }
        return true;
    }

    private void endQuiz() {
        int score = 0;
        for (int i = 0; i < 10; i++) {
            if (cAnswers[i] == questions.get(i).getCorrect_answer_id())
                score++;
        }

        Intent intent = new Intent(this, QuizResultActivity.class);
        intent.putExtra("cAnswers", cAnswers);
        intent.putExtra("score", score);
        intent.putExtra("questions", (Serializable) questions);
        startActivity(intent);

        finish();
    }

    private List<QuizQuestionsDB> getQuestions() {
        return db.quizQuestionDao().getAll();
    }

    private List<QuizAnswersDB> getAnswers(int qId) {
        return db.quizAnswerDao().getAnswers(qId);
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/QuizActivity.kt
package bassamalim.hidaya.activities

import android.content.Intent
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import bassamalim.hidaya.R
import bassamalim.hidaya.database.AppDatabase
import bassamalim.hidaya.database.dbs.QuizQuestionsDB
import bassamalim.hidaya.ui.components.RadioGroup
import bassamalim.hidaya.ui.components.MyButton
import bassamalim.hidaya.ui.components.MyScaffold
import bassamalim.hidaya.ui.components.MyText
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.ui.theme.Grey
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.DBUtils
import java.io.Serializable
import java.util.*

class QuizActivity : ComponentActivity() {

    private lateinit var db: AppDatabase
    private lateinit var questions: List<QuizQuestionsDB>
    private val cAnswers = IntArray(10)
    private val current = mutableStateOf(0)
    private val currentAs = mutableStateListOf("", "", "", "")
    private val selection = mutableStateOf(-1)
    private val allAnswered = mutableStateOf(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ActivityUtils.onActivityCreateSetLocale(this)

        db = DBUtils.getDB(this)

        getQuestions()

        Arrays.fill(cAnswers, -1)

        ask(current.value)

        setContent {
            AppTheme {
                UI()
            }
        }
    }

    private fun getQuestions() {
        val rawQuestions = db.quizQuestionDao().all.toMutableList()
        rawQuestions.shuffle()
        questions = ArrayList(rawQuestions.subList(0, 10))
    }

    private fun ask(num: Int) {
        current.value = num

        val q = questions[num]
        val answers = db.quizAnswerDao().getAnswers(q.getQuestionId())
        for (i in currentAs.indices) currentAs[i] = answers[i].answer_text!!

        selection.value = cAnswers[current.value]
    }

    private fun answered(a: Int) {
        cAnswers[current.value] = a

        allAnswered.value = allAnswered()
        if (current.value != 9) nextQ()
    }

    private fun nextQ() {
        if (current.value == 9) endQuiz()
        else ask(++current.value)
    }

    private fun previousQ() {
        if (current.value > 0) ask(--current.value)
    }

    private fun allAnswered(): Boolean {
        for (cAnswer in cAnswers) {
            if (cAnswer == -1) return false
        }
        return true
    }

    private fun endQuiz() {
        var score = 0
        for (i in 0..9) {
            if (cAnswers[i] == questions[i].getCorrectAnswerId()) score++
        }

        val intent = Intent(this, QuizResultActivity::class.java)
        intent.putExtra("cAnswers", cAnswers)
        intent.putExtra("score", score)
        intent.putExtra("questions", questions as Serializable?)
        startActivity(intent)

        finish()
    }

    @Composable
    private fun UI() {
        MyScaffold("${getString(R.string.question)} ${current.value + 1}") {
            Column(
                Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.SpaceBetween,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Box(
                    Modifier
                        .fillMaxWidth()
                        .heightIn(1.dp, 200.dp)
                        .verticalScroll(rememberScrollState())
                        .background(AppTheme.colors.primary),
                    contentAlignment = Alignment.Center
                ) {
                    MyText(
                        text = questions[current.value].getQuestionText(),
                        fontSize = 28.sp,
                        textColor = AppTheme.colors.onPrimary,
                        modifier = Modifier.padding(vertical = 10.dp, horizontal = 10.dp)
                    )
                }

                RadioGroup(
                    options = currentAs,
                    selection = selection,
                    modifier = Modifier
                        .heightIn(1.dp, 400.dp)
                        .verticalScroll(rememberScrollState())
                ) { index ->
                    answered(index)
                }

                Row(
                    Modifier
                        .fillMaxWidth()
                        .weight(1F, false)
                        .padding(bottom = 20.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.SpaceAround
                ) {
                    MyButton(
                        text = stringResource(R.string.previous_question),
                        textColor =
                            if (current.value == 0) Grey
                            else AppTheme.colors.text,
                        innerPadding = PaddingValues(10.dp)
                    ) {
                        if (current.value != 0) previousQ()
                    }

                    MyButton(
                        text = stringResource(
                            if (current.value == 9) {
                                if (allAnswered.value) R.string.finish_quiz
                                else R.string.answer_all_questions
                            }
                            else R.string.next_question
                        ),
                        modifier = Modifier.sizeIn(maxWidth = 175.dp),
                        textColor =
                            if (current.value == 9) {
                                if (allAnswered.value) AppTheme.colors.text
                                else Grey
                            }
                            else AppTheme.colors.text,
                        innerPadding = PaddingValues(10.dp)
                    ) {
                        if (current.value == 9) {
                            if (allAnswered.value) nextQ()
                        }
                        else nextQ()
                    }
                }
            }
        }
    }

}